lib/trino-filesystem/src/main/java/io/trino/filesystem/memory/MemoryFileSystemCache.java

```diff
@@ -0,0 +1,189 @@
+    }
+
+    public MemoryFileSystemCache(Tracer tracer, Duration expireAfterWrite, DataSize maxSize, DataSize maxContentLength)
+    {
+        this.tracer = requireNonNull(tracer, "tracer is null");
+        checkArgument(maxContentLength.compareTo(DataSize.of(1, GIGABYTE)) <= 0, "maxContentLength must be less than or equal to 1GB");
+        this.cache = EvictableCacheBuilder.newBuilder()
+                .maximumWeight(maxSize.toBytes())
+                .weigher((Weigher<CacheKey, Slice>) (key, value) -> toIntExact(estimatedSizeOf(key.providedKey()) + Integer.BYTES + value.getRetainedSize()))
+                .expireAfterWrite(expireAfterWrite.toMillis(), TimeUnit.MILLISECONDS)
+                .shareNothingWhenDisabled()
+                .recordStats()
+                .build();
+        this.maxContentLengthBytes = toIntExact(maxContentLength.toBytes());
+    }
+
+    @Override
+    public TrinoInput cacheInput(TrinoInputFile delegate, String key)
+            throws IOException
+    {
+        long fileSize = delegate.length();
+        if (fileSize > maxContentLengthBytes) {
+            return delegate.newInput();
+        }
+
+        Span span = createSpan(key, delegate.location(), fileSize, "cacheInput");
+        return withTracing(span, () -> new MemoryInput(delegate.location(), getFromCache(key, delegate)));
+    }
+
+    @Override
+    public TrinoInputStream cacheStream(TrinoInputFile delegate, String key)
+            throws IOException
+    {
+        long fileSize = delegate.length();
+        if (fileSize > maxContentLengthBytes) {
+            return delegate.newStream();
+        }
+
+        Span span = createSpan(key, delegate.location(), fileSize, "cacheInputStream");
+        return withTracing(span, () -> new MemoryInputStream(delegate.location(), getFromCache(key, delegate)));
+    }
+
+    @Override
+    public void expire(Location location)
+            throws IOException
+    {
+        List<CacheKey> expired = cache.asMap().keySet().stream()
+                .filter(key -> key.providedKey().contains(location.path()))
+                .collect(toImmutableList());
+        cache.invalidateAll(expired);
+    }
+
+    @Managed
+    public void flushCache()
+    {
+        cache.invalidateAll();
+    }
+
+    @Managed
+    public long getHitCount()
+    {
+        return cache.stats().hitCount();
+    }
+
+    @Managed
+    public long getRequestCount()
+    {
+        return cache.stats().requestCount();
+    }
+
+    @VisibleForTesting
+    boolean isCached(String key, long fileSize)
+    {
+        return cache.getIfPresent(new CacheKey(key, toIntExact(fileSize))) != null;
+    }
+
+    private Slice getFromCache(String key, TrinoInputFile delegate)
+            throws IOException
+    {
+        try {
+            // File length is available for free at this point, we add it to cache key to be resilient to manual modifications to file
```

Comment:
If we have file length, then we would also have last modified and could use that instead? (assuming Delta. With Iceberg it's probably safe to assume that the files are immutable).

Reply:
Iceberg metadata gives length for free, but not file modified time as far as I can tell